parsons/targetsmart/targetsmart_smartmatch.py

```diff
@@ -0,0 +1,311 @@
+
+    return input_table.addrownumbers(field=INTERNAL_JOIN_ID)
+
+
+def _prepare_input(intable, tmpdir):
+    valid = VALID_FIELDS + [INTERNAL_JOIN_ID]
+    supported = set(intable.fieldnames()) & set(valid)
+    if not supported:
+        raise SmartMatchError(
+            "No supported field identifiers were found in the input table."
+            f" Expecting one or more from: {VALID_FIELDS}"
+        )
+    return intable.cut(*supported)
+
+
+class SmartMatch:
+    """
+    Works as a mixin to the TargetSmartAPI class.
+    """
+
+    def __init__(self):
+        # Set by TargetSmartAPI constructor
+        self.connection = None
+
+    def _smartmatch_poll(self, poll_url, submit_filename):
+        download_url = None
+        while True:
+            poll_response = requests.get(
+                poll_url,
+                {"filename": submit_filename},
+                headers=self.connection.headers,
+            )
+
+            if poll_response.ok:
+                poll_info = poll_response.json()
+
+                if poll_info["error"]:
+                    raise SmartMatchError(poll_info["error"])
+
+                download_url = poll_info["url"]
+                if download_url:
+                    break
+            time.sleep(60 * 2.5)
+        return download_url
+
+    def smartmatch(
+        self,
+        input_table,
+        max_matches=1,
+        include_email=False,
+        include_landline=False,
+        include_wireless=False,
+        include_voip=False,
+        tmp_location=None,
+        keep_smartmatch_input_file=False,
+        keep_smartmatch_output_gz_file=False,
+    ):
+        """Submit the contact list records available in the Parsons table ``input_table`` to
+        TargetSmart SmartMatch.
+
+        * `SmartMatch overview <https://docs.targetsmart.com/my_tsmart/smartmatch/overview.html>`_
+        * `SmartMatch API doc <https://docs.targetsmart.com/developers/tsapis/v2/service/smartmatch.html>`_
+        * `Supported input header field identifiers <https://docs.targetsmart.com/developers/tsapis/v2/service/smartmatch.html#supported-field-identifiers>`_
+
+        Your application provides a contact list which will be matched to
+        TargetSmart’s database of voting age individuals.
+
+        `TargetSmart Client Services <mailto:support@targetsmart.com>`_
+        provisions SmartMatch for your API key, configuring the fields from the
+        TargetSmart Data Dictionary that will be appended to each matched
+        record.
+
+        This method blocks until TargetSmart has completed the remote workflow
+        execution. The execution time can take minutes to hours to complete
+        depending on the file size, the types of field identifiers present, and
+        TargetSmart system load. SmartMatch executions cannot be canceled once
+        submitted to TargetSmart.
+
+        Since Parsons Petl tables are lazy, the SmartMatch output file is always
```

Comment:
Since Parsons tables are lazy.

* I think that most users don't know what petl, so just striking for the purposes of clarity.